ehrql/dummy_data_nextgen/generator.py

```diff
@@ -106,6 +111,25 @@ def get_data(self):
                 if found >= self.population_size:
                     break
 
+            if generator.required_tables is None and valid_patient_ids:
```

Comment:
Just trying to understand what's going on here; in the first batch of patients generated, we go through all of the tables in the database, and built a dict which has keys patient_id and values a set of tables that that patient appears in. Then we loop through all the patient IDs and update `required_tables` and `forbidden_tables` to be all tables that at least one patient has an entry in, and all tables that no patients have any entries in respectively.

And then on line 324 we use the existence of the required tables built in this step, so on subsequent batches of generated patient IDs, if the table is in required tables, the row count for event-level tables is increased by 1.

Tables in the database are populated by the tables in the population query (other tables that are needed for the dataset by aren't required for determining population membership are added later, but not added to the database), so if a tables isn't required for the population query, it won't be in the required tables and it will have a lower event-level row count.

I'm struggling to understand why we're doing this though.

Reply:
Actually, maybe I do understand. The geometric distribution can be 0, so if the table is required for the population inclusion, it ensures that the rows are at least 1.

But this isn't the case on the first batch, is it? Because then self.required_tables is None, so if the row count is calculated as 0, it won't be incremented.